File: conformance/bundles/02-context-isolation/counter.steps.rs
//! Rust sibling of `counter.steps.ts` (bundle `02-context-isolation`).

use std::collections::BTreeMap;
use varar::{HandlerError, Registry, Steps, Value};

fn count_of(state: &Value) -> i64 {
    match state {
        Value::Map(m) => match m.get("count") {
            Some(Value::Int(i)) => *i,
            _ => 0,
        },
        _ => 0,
    }
}

pub fn register(r: Registry) -> Registry {
    let mut s = Steps::from_registry(r);
    s.stimulus("I increment", |state| {
        let next = count_of(&state) + 1;
        Ok(Some(Value::Map(BTreeMap::from([(
            "count".to_string(),
            Value::Int(next),
        )]))))
    });
    s.sensor("The count is {int}", |state, n| {
        let count = count_of(&state);
        let expected = if let Value::Int(i) = n { i } else { 0 };
        if count != expected {
            return Err(HandlerError::new(format!(
                "expected {expected} but got {count}"
            )));
        }
        Ok(None)
    });
    s.into_registry()
}

pub fn state() -> Value {
    Value::Map(BTreeMap::from([("count".to_string(), Value::Int(0))]))
}

File: conformance/bundles/02-context-isolation/counter.steps.go
// Go sibling of counter.steps.ts (bundle 02-context-isolation).
package fixture

import "github.com/varar-dev/varar-go/varar"

func countOf(state varar.Value) int {
	if m, ok := state.AsMap(); ok {
		if c, ok := m["count"]; ok {
			if n, ok := c.AsInt(); ok {
				return int(n)
			}
		}
	}
	return 0
}

func Register(s *varar.Steps) {
	s.Stimulus("I increment", func(state varar.Value) (varar.Value, error) {
		return varar.MapValue(map[string]varar.Value{"count": varar.IntValue(int64(countOf(state) + 1))}), nil
	})
	// One slot ({int}): return the observed count and let the core compare it
	// against the number in the document, rather than asserting by hand.
	s.Sensor("The count is {int}", func(state varar.Value, expected int) (int, error) {
		return countOf(state), nil
	})
}

func State() varar.Value {
	return varar.MapValue(map[string]varar.Value{"count": varar.IntValue(0)})
}
